File: src/RealWorld.Core/Entities/Article.cs
using System.Text.RegularExpressions;

namespace RealWorld.Core.Entities;

public class Article
{
    public string Id { get; set; } = string.Empty;
    public string UserId { get; set; } = string.Empty;
    public string Slug { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public string Body { get; set; } = string.Empty;
    public List<Tag> Tags { get; set; } = new();
    public DateTimeOffset CreatedAt { get; set; }
    public DateTimeOffset UpdatedAt { get; set; }

    public Article() { }

    public Article(string title, string description, string body, List<string> tagList, string userId)
        : this(title, description, body, tagList, userId, DateTimeOffset.UtcNow)
    {
    }

    public Article(string title, string description, string body, List<string> tagList, string userId, DateTimeOffset createdAt)
    {
        Id = Guid.NewGuid().ToString();
        Slug = ToSlug(title);
        Title = title;
        Description = description;
        Body = body;
        Tags = tagList.Distinct().Select(t => new Tag(t)).ToList();
        UserId = userId;
        CreatedAt = createdAt;
        UpdatedAt = createdAt;
    }

    public void Update(string? title, string? description, string? body)
    {
        if (!string.IsNullOrWhiteSpace(title))
        {
            Title = title;
            Slug = ToSlug(title);
            UpdatedAt = DateTimeOffset.UtcNow;
        }
        if (!string.IsNullOrWhiteSpace(description))
        {
            Description = description;
            UpdatedAt = DateTimeOffset.UtcNow;
        }
        if (!string.IsNullOrWhiteSpace(body))
        {
            Body = body;
            UpdatedAt = DateTimeOffset.UtcNow;
        }
    }

    public static string ToSlug(string title)
    {
        return Regex.Replace(title.ToLowerInvariant(), @"[&\uFE30-\uFFA0'\""\s?,\.]+", "-");
    }

    public override bool Equals(object? obj) => obj is Article a && a.Id == Id;
    public override int GetHashCode() => Id.GetHashCode();
}

File: services/article-service/src/main/java/io/spring/articleservice/core/article/Article.java
package io.spring.articleservice.core.article;

import io.spring.shared.util.Util;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.joda.time.DateTime;

@Getter
@NoArgsConstructor
@EqualsAndHashCode(of = {"id"})
public class Article {
  private String userId;
  private String id;
  private String slug;
  private String title;
  private String description;
  private String body;
  private List<Tag> tags;
  private DateTime createdAt;
  private DateTime updatedAt;

  public Article(
      String title, String description, String body, String[] tagList, String userId) {
    this(title, description, body, tagList, userId, new DateTime());
  }

  public Article(
      String title,
      String description,
      String body,
      String[] tagList,
      String userId,
      DateTime createdAt) {
    this.id = UUID.randomUUID().toString();
    this.slug = toSlug(title);
    this.title = title;
    this.description = description;
    this.body = body;
    this.tags =
        Arrays.stream(tagList).collect(Collectors.toList()).stream()
            .map(Tag::new)
            .collect(Collectors.toList());
    this.userId = userId;
    this.createdAt = createdAt;
    this.updatedAt = createdAt;
  }

  public void update(String title, String description, String body) {
    if (!Util.isEmpty(title)) {
      this.title = title;
      this.slug = toSlug(title);
    }
    if (!Util.isEmpty(description)) {
      this.description = description;
    }
    if (!Util.isEmpty(body)) {
      this.body = body;
    }
    this.updatedAt = new DateTime();
  }

  public static String toSlug(String title) {
    return title.toLowerCase().replaceAll("[&|\\uFE30-\\uFFA0'\"\\s?,\\.]+", "-");
  }
}
